File: src/1357-Apply-Discount-Every-n-Orders/1357.js
/**
 * @param {number} n
 * @param {number} discount
 * @param {number[]} products
 * @param {number[]} prices
 */
var Cashier = function(n, discount, products, prices) {
    this.n = n, this.discount = discount, this.cnt = 0;
    this.price = new Map();
    for (let i = 0; i < products.length; ++i)
        this.price.set(products[i], prices[i]);
};

/** 
 * @param {number[]} product 
 * @param {number[]} amount
 * @return {number}
 */
Cashier.prototype.getBill = function(product, amount) {
    let total = 0;
    for (let i = 0; i < product.length; ++i)
        total += this.price.get(product[i]) * amount[i];
    return total * (++this.cnt % this.n == 0 ? 1 - this.discount / 100.0 : 1);
};

File: src/1357-Apply-Discount-Every-n-Orders/1357.cpp
class Cashier 
{
public:
    Cashier(int n, int discount, vector<int>& products, vector<int>& prices) 
    {
        for (int i = 0; i < products.size(); i++) price[products[i]] = prices[i];
        this->n = n, this->discount = discount;
    }
    
    double getBill(vector<int> product, vector<int> amount) 
    {
        this->cnt++;
        double total = 0;
        
        for (int i = 0; i < product.size(); i++) total += price[product[i]] * amount[i];
        return cnt % n ? total : total * (1 - discount/100.0);
    }
private:
    int n, discount, cnt = 0;  
    unordered_map<int, int> price;
};